module/x/gravity/keeper/grpc_query.go

```diff
@@ -1,354 +1,407 @@
+	if !ok {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to signer set for %d", req.Nonce)
+	}
+
+	// TODO: ensure that latest signer set tx nonce index is set properly
+	// TODO: ensure nonce sequence starts at one
+
+	return &types.SignerSetTxResponse{SignerSet: ss}, nil
+}
+
+func (k Keeper) BatchTx(c context.Context, req *types.BatchTxRequest) (*types.BatchTxResponse, error) {
+	if !common.IsHexAddress(req.ContractAddress) {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "invalid hex address %s", req.ContractAddress)
+	}
+	ctx := sdk.UnwrapSDKContext(c)
+	res := &types.BatchTxResponse{}
+
+	if req.Nonce == 0 {
+		store := prefix.NewStore(ctx.KVStore(k.storeKey), append([]byte{types.OutgoingTxKey}, types.BatchTxPrefixByte))
+		iter := store.ReverseIterator(nil, nil)
+		defer iter.Close()
+		for ; iter.Valid(); iter.Next() {
+			var any cdctypes.Any
+			k.cdc.MustUnmarshalBinaryBare(iter.Value(), &any)
+			var otx types.OutgoingTx
+			if err := k.cdc.UnpackAny(&any, &otx); err != nil {
+				panic(err)
+			}
+			batch, ok := otx.(*types.BatchTx)
+			if !ok {
+				return nil, sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to batch tx for %d %s", req.Nonce, req.ContractAddress)
+			}
+			if batch.TokenContract == req.ContractAddress {
+				res.Batch = batch
+				break
+			}
+		}
+	} else {
+		// TODO: audit once we finalize storage
+		storeIndex := append(sdk.Uint64ToBigEndian(req.Nonce), common.Hex2Bytes(req.ContractAddress)...)
+		otx := k.GetOutgoingTx(sdk.UnwrapSDKContext(c), types.MakeOutgoingTxKey(storeIndex))
+		if otx == nil {
+			return nil, sdkerrors.Wrapf(types.ErrInvalid, "no batch tx found for %d %s", req.Nonce, req.ContractAddress)
+		}
+		batch, ok := otx.(*types.BatchTx)
+		if !ok {
+			return nil, sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to batch tx for %d %s", req.Nonce, req.ContractAddress)
+		}
+		res.Batch = batch
+	}
+
+	return res, nil
+}
+
+func (k Keeper) ContractCallTx(c context.Context, req *types.ContractCallTxRequest) (*types.ContractCallTxResponse, error) {
+	storeIndex := append(sdk.Uint64ToBigEndian(req.InvalidationNonce), req.InvalidationScope...)
+	otx := k.GetOutgoingTx(sdk.UnwrapSDKContext(c), types.MakeOutgoingTxKey(storeIndex))
+	if otx == nil {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "no contract call found for %d %s", req.InvalidationNonce, req.InvalidationScope)
+	}
+
+	cctx, ok := otx.(*types.ContractCallTx)
+	if !ok {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to contract call for %d %s", req.InvalidationNonce, req.InvalidationScope)
+	}
+
+	// TODO: figure out how to call latest
+
+	return &types.ContractCallTxResponse{LogicCall: cctx}, nil
+}
+
+func (k Keeper) SignerSetTxs(c context.Context, req *types.SignerSetTxsRequest) (*types.SignerSetTxsResponse, error) {
+	var signers []*types.SignerSetTx
+	k.IterateOutgoingTxsByType(sdk.UnwrapSDKContext(c), types.SignerSetTxPrefixByte, func(_ []byte, otx types.OutgoingTx) bool {
```

Comment:
use query pagination. Check SDK for reference

Reply:
Done